README.rst
=====================
Celery Redis Sentinel
=====================

.. image:: https://badge.fury.io/py/celery-redis-sentinel.svg
    :target: http://badge.fury.io/py/celery-redis-sentinel

.. image:: https://travis-ci.org/dealertrack/celery-redis-sentinel.svg?branch=master
    :target: https://travis-ci.org/dealertrack/celery-redis-sentinel

.. image:: https://coveralls.io/repos/dealertrack/celery-redis-sentinel/badge.svg?branch=master&service=github
    :target: https://coveralls.io/github/dealertrack/celery-redis-sentinel?branch=master


Celery broker and results backend implementation for
`Redis Sentinel <http://redis.io/topics/sentinel>`_

* Free software: MIT license
* GitHub: https://github.com/dealertrack/celery-redis-sentinel
* Documentation: https://celery-redis-sentinel.readthedocs.org.

Why?
----

`Redis <http://redis.io/>`_ is a pretty awesome in-memory key-value data-store.
Being in-memory makes it wickedly fast however at a cost of no-persistence.
In business-critical applications (you know, which make company money) that makes
stand-alone redis deployment non-practical. This is where
`Redis Sentinel <http://redis.io/topics/sentinel>`_ comes in.
It provides scalability and high availability to Redis 2.X
(Redis 3.X comes with native-clustering which is different from Sentinel).
As a result, Redis becomes a viable solution for solving some of business needs.
As you can imagine from the project title, one use-case is using Redis Sentinel with
`celery <http://www.celeryproject.org/>`_.
Unfortunately celery does not support Redis Sentinel by default hence this
library which aims to provide non-official Redis Sentinel support as **both**
celery broker and results backend.

Installing
----------

Installation is super easy with ``pip``::

    $ pip install celery-redis-sentinel

Usage
-----

Using this library is pretty simple. All you need to do is configure celery
to use Redis Sentinel for broker and/or results backend. That is done
with a couple of settings::

    # celeryconfig.py
    BROKER_URL = 'redis-sentinel://redis-sentinel:26379/0'
    BROKER_TRANSPORT_OPTIONS = {
        'sentinels': [('192.168.1.1', 26379),
                      ('192.168.1.2', 26379),
                      ('192.168.1.3', 26379)],
        'service_name': 'master',
        'socket_timeout': 0.1,
    }

    CELERY_RESULT_BACKEND = 'redis-sentinel://redis-sentinel:26379/1'
    CELERY_RESULT_BACKEND_TRANSPORT_OPTIONS = BROKER_TRANSPORT_OPTIONS

Some notes about the configuration:

* note the use of ``redis-sentinel`` schema within the URL for broker and results
  backend. In order to use that schema, which is not shipped with celery, where you create
  your celery application you **must** first need to register sentinel support::

      # tasks.py
      from celery_redis_sentinel import register

      # has to be called before creating celery app
      register()

      app = Celery('tasks')
* hostname and port are ignored within the actual URL. Sentinel uses transport options
  ``sentinels`` setting to create a ``Sentinel()`` instead of configuration URL.

Scheduling During Failover
--------------------------

Some considerations while using Redis Sentinel as a celery broker. While the failover
is in progress, no tasks can be scheduled. Trying to schedule a task will either
raise ``Timeout`` or ``ConnectionError``. That is because other sentinel nodes
within the cluster, depending on the configuration, have a timeout until they elect
a new master. During that time, trying to schedule a task will attempt to store
it in now-invalid master node hence the exception. If that is unacceptable within
your application, this library comes with a small wrapper which allows to trigger
tasks which will block the scheduling until new master will be elected::

    from celery_redis_sentinel.redis_sentinel import ensure_redis_call
    from tasks import add

    # this will blow up during failover
    add.delay(1, 2)
    # this will keep retrying until it succeeds
    ensure_redis_call(add.delay, 1, 2)

Alternatively you can use a supplied abstract celery task subclass which provides
same retrying behavior in the task definition itself::

    # tasks.py
    from celery_redis_sentinel.task import EnsuredRedisTask

    @app.task(base=EnsuredRedisTask)
    def add(a, b):
        return a + b


HISTORY.rst
.. :changelog:

History
-------

Master (not yet on PyPI)
~~~~~~~~~~~~~~~~~~~~~~~~

In Progress

0.3 (2016-05-03)
~~~~~~~~~~~~~~~~

* **New**: Addition of ``ShortLivedStrictRedis`` and ``ShortLivedSentinel``.
  Both of them use short-lived connections which disconnect from redis
  as soon as the query to redis is complete.
* **Fixed**: All sentinel connections are now created via ``ShortLivedSentinel``.
  This fixes an issue when sentinel would reach its max connections limit
  since all celery workers would always be connected to sentinel.
  That is not necessary since sentinel is queried very rarely for the current
  master connection details.
  In addition this is useful when Redis Sentinel is used behind a firewall
  since sentinel would not notice when firewall would close the connections
  and so would not release them.

0.2 (2016-01-14)
~~~~~~~~~~~~~~~~

* **New**: Added ``EnsuredRedisTask`` which allows to ensure tasks are scheduled
  via an abstract base task class in task definition rather then explicitly using
  ``ensure_redis_call`` while calling the task::

      @app.task(base=EnsuredRedisTask)
      def foo(): pass

0.1 (2016-01-13)
~~~~~~~~~~~~~~~~

* First release


celery_redis_sentinel/backend.py
# -*- coding: utf-8 -*-
from __future__ import absolute_import, print_function, unicode_literals

from celery.backends.redis import RedisBackend
from kombu.utils import cached_property
from redis import Redis

from .redis_sentinel import EnsuredRedisMixin, get_redis_via_sentinel


class RedisSentinelBackend(RedisBackend):
    """
    Redis results backend with support for Redis Sentinel

    .. note::
        In order to correctly configure the sentinel,
        this backend expects an additional backend celery
        configuration to be present - ``CELERY_RESULT_BACKEND_TRANSPORT_OPTIONS``.
        Here is are sample transport options::

            CELERY_RESULT_BACKEND_TRANSPORT_OPTIONS = {
                'sentinels': [('192.168.1.1', 26379),
                              ('192.168.1.2', 26379),
                              ('192.168.1.3', 26379)],
                'service_name': 'master',
                'socket_timeout': 0.1,
            }
    """

    def __init__(self, transport_options=None, *args, **kwargs):
        super(RedisSentinelBackend, self).__init__(*args, **kwargs)

        _get = self.app.conf.get
        self.transport_options = transport_options or _get('CELERY_RESULT_BACKEND_TRANSPORT_OPTIONS') or {}
        self.sentinels = self.transport_options['sentinels']
        self.service_name = self.transport_options['service_name']
        self.socket_timeout = self.transport_options.get('socket_timeout', 0.1)

    @cached_property
    def client(self):
        """
        Cached property for getting ``Redis`` client to be used to interact with redis.

        Returned client also subclasses from :class:`.EnsuredRedisMixin` which
        ensures that all redis commands are executed with retry logic in case
        of sentinel failover.

        Returns
        -------
        Redis
            Redis client connected to Sentinel via Sentinel connection pool
        """
        params = self.connparams
        params.update({
            'sentinels': self.sentinels,
            'service_name': self.service_name,
            'socket_timeout': self.socket_timeout,
        })
        return get_redis_via_sentinel(
            redis_class=type(str('Redis'), (EnsuredRedisMixin, Redis), {}),
            **params
        )
